beacon_node/eth1_chain/src/web3_fetcher.rs

```diff
@@ -0,0 +1,245 @@
+            "get_hash_tree_root",
+            (),
+            None,
+            Options::default(),
+            block_number,
+        );
+        let deposit_root: Vec<u8> = data.wait().ok()?;
+        Some(H256::from_slice(&deposit_root))
+    }
+
+    /// Returns a future which subscribes to `DepositEvent` events and inserts the
+    /// parsed deposit into the passed cache structure everytime an event is emitted.
+    fn get_deposit_logs_subscription(
+        &self,
+        cache: Arc<RwLock<BTreeMap<u64, DepositData>>>,
+    ) -> Box<Future<Item = (), Error = ()> + Send> {
+        let filter: Filter = self.get_deposit_logs_filter();
+        let event_future = self
+            .web3
+            .eth_subscribe()
+            .subscribe_logs(filter)
+            .then(move |sub| {
+                sub.unwrap().for_each(move |log| {
+                    let parsed_logs = parse_deposit_logs(log).unwrap();
+                    let mut logs = cache.write();
+                    logs.insert(parsed_logs.0, parsed_logs.1);
+                    Ok(())
+                })
+            })
+            .map_err(|_| ());
+        Box::new(event_future)
+    }
+}
+
+// Converts a valid vector to a u64.
+pub fn vec_to_u64_le(bytes: &[u8]) -> Option<u64> {
+    let mut array = [0; 8];
+    if bytes.len() == 8 {
+        let bytes = &bytes[..array.len()];
+        array.copy_from_slice(bytes);
+        Some(u64::from_le_bytes(array))
+    } else {
+        None
+    }
+}
+
+/// Parse contract logs.
+pub fn parse_logs(log: Log, types: &[ParamType]) -> Option<Vec<Token>> {
+    decode(types, &log.data.0).ok()
+}
+
+/// Parse logs from deposit contract.
+pub fn parse_deposit_logs(log: Log) -> Option<(u64, DepositData)> {
+    let deposit_event_params = &[
+        ParamType::FixedBytes(48), // pubkey
+        ParamType::FixedBytes(32), // withdrawal_credentials
+        ParamType::FixedBytes(8),  // amount
+        ParamType::FixedBytes(96), // signature
+        ParamType::FixedBytes(8),  // index
+    ];
+    let parsed_logs = parse_logs(log, deposit_event_params).unwrap();
+    // Convert from tokens to Vec<u8>.
+    let params = parsed_logs
+        .into_iter()
+        .map(|x| match x {
+            Token::FixedBytes(v) => Some(v),
+            _ => None,
+        })
+        .collect::<Option<Vec<_>>>()?;
+
+    // Event should have exactly 5 parameters.
+    if params.len() == 5 {
+        Some((
+            vec_to_u64_le(&params[4])?,
+            DepositData {
+                pubkey: PublicKeyBytes::from_bytes(&params[0]).unwrap(),
+                withdrawal_credentials: H256::from_slice(&params[1]),
+                amount: vec_to_u64_le(&params[2])?,
+                signature: SignatureBytes::from_bytes(&params[3]).ok()?,
+            },
+        ))
+    } else {
+        None
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    // Note: Running tests using ganache-cli instance with config
```

Comment:
We'll have to figure out how to deal with integration tests for this.

Our CI server will likely fail these because it's not running ganache.